File: WebSite/EulerAList/pset21.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <climits>
#include <cmath>
#include <unordered_map>
using namespace std;

vector<int> primeList{2, 3, 5, 7, 11};
vector<long> d(10000, 0);

long findDivisorSum(int value)
{
    int tmp = value;
    long result = 1;
    for (int i : primeList)
    {
        if (value < i || value <= 1) break;
        int cnt = 0;
        while (value % i == 0 && value > 1)
        {
            value /= i;
            cnt ++;
        }
        result *= (long)(pow(i,cnt + 1) - 1) / (i - 1);
    }
    if (value > 1)
    {
        primeList.push_back(value);
        result *= value + 1;
    }
    result -= tmp;
    return result;
}

int main()
{   
    long result = 0;
    for (int i = 2; i < 10000; i ++)
    {
        d[i] = findDivisorSum(i);
    }
    for (int a = 2; a < 10000; a ++)
    {
        if (d[d[a]] == a && d[a] != a)
        {
            cout << a << ";";
            result += a;
        }
    }
    cout << endl;
    cout << result << endl;
    return 0;
}

File: WebSite/EulerAList/pset21.py
primeList = [2, 3, 5, 7, 11]
d = [0] * 10000

def findDivisorSum(value):
    tmp = value
    result = 1
    for i in primeList:
        if value < i or value <= 1: 
            break
        cnt = 0
        while value % i == 0 and value > 1:
            value /= i
            cnt += 1
        result *= int(pow(i,cnt + 1) - 1) // (i - 1)
    if value > 1:
        primeList.append(value)
        result *= value + 1
    result -= tmp
    return result

result = 0
for i in range(2, 10000):
    d[i] = findDivisorSum(i)
for a in range(2, 10000):
    if d[a] < 10000 and d[d[a]] == a and d[a] != a:
        result += a
print(result)